### DISCUSSION TITLE
Proposal: Allow declaring local variables with "out var" to make them continue to be accessible outside of a block

### DISCUSSION THREAD

**Discussion by slimCODE** [dotnet/csharplang#1075] [General]
Scenario: When using the `using(IDisposable)` pattern to acquire something, you often need to declare variables right before, forcing you to specify the type and initialize it.

```csharp
ISomething thing = null;

using(var disposable = GetDisposable())
{
    thing = disposable.GetThing();
}

thing.Consume();
```

With the following construct, you would save one line, and avoid having to describe the variable type.

```csharp
using(var disposable = GetDisposable())
{
    out var thing = disposable.GetThing();
}

thing.Consume();
```

**Comment by alrz** [comment]
relates to https://github.com/dotnet/csharplang/issues/451

**Comment by svick** [comment]
Small correction: you don't have to initialize the variable, this is valid code:

```c#
ISomething thing;

using(var disposable = GetDisposable())
{
    thing = disposable.GetThing();
}

thing.Consume();
```

**Comment by jcouv** [comment]
This doesn't appeal to me. It's unclear to which scope the declaration should escape. Also, I don't feel this helps readability.

**Comment by bartdesmet** [comment]
A subset of the problems this is trying to solve may be solvable by introducing `using` expressions:

```csharp
string text = using (var sr = new StreamReader(path)) => sr.ReadToEnd();
```

where I've taken the liberty to use the `=>` syntax for the first time to "promote" a statement to an expression, given that single-body-statement `using` statements are already valid. Syntax could be debated at length, e.g. if `=>` is needed at all. The bigger question is whether "expression everything" is warranted. I could see it being useful for `switch` (already being discussed elsewhere) and `lock`:

```csharp
string five = lock (gate) => dictionary[5];
```
